### PR TITLE
MinCut pooling for sparse adjacency matrix

### PR DESCRIPTION
**Pull request by xiaohan2012**
# Motivation

The current `nn.dense.mincut_pool` requires the input adjacency matrix to be a dense matrix. This requirement does not scale to large graphs

# What

This PR considers a sparse counterpart of `nn.dense.mincut_pool`, which takes in sparse adjacency matrix as input.

# Speed comparison with dense mincut

- Benchmarked on a 8k x 8k adjacency matrix of sparsity ratio 0.01
- Device: CPU
- OS: OSX 14.2.1 (M2 chip)

<details>
<summary>The script</summary>

```python
import torch
from contexttimer import Timer
from torch_geometric.nn.pool.mincut import mincut_pool as mincut_pool_sparse
from torch_geometric.nn.dense.mincut_pool import  dense_mincut_pool


batch_size, num_nodes, channels, num_clusters = (10, 8000, 100, 10)
sparsity = 0.01

# batched feature matrix
x = torch.randn((batch_size, num_nodes, channels))
adj_dense = (torch.rand((batch_size, num_nodes, num_nodes)) <= sparsity).type(
    torch.float
)
adj_sparse = adj_dense.to_sparse()
# batched node clustering tensor
s = torch.randn((batch_size, num_nodes, num_clusters))
# batched random masks
mask = torch.randint(0, 2, (batch_size, num_nodes), dtype=torch.bool)

with Timer() as t:
    mincut_pool_sparse(x, adj_sparse, s, mask)
    print(f"Execution time: {t.elapsed} seconds")

with Timer() as t:    
    dense_mincut_pool(x, adj_dense, s, mask)
    print(f"Execution time: {t.elapsed} seconds")
```
</details>

```
Spase mincut pooling: 0.5210544160217978 s
Dense mincut pooling: 1.2443664169986732 s
```
# Remark on matrix multiplication

`torch.sparse.mm` is used, therefore restricting the use to COO matrices only. As a future work, multiplying CSR matrix with CSC matrix should be used for better performance

### CODE REVIEW

**Comment by xiaohan2012** [comment]
@rusty1s is it possible to ask for your review?

**Comment by xiaohan2012** [comment]
@wsad1 @EdisonLeeeee @rusty1s can I ask for a review?

**Comment by carloabate** [comment]
Hey! We developed `tgp` (`torch-geometric-pool`), a graph pooling library for GNNs that operates on top of PyG. In particular, we implemented `MinCutPooling`, including a sparse adjacency path: in `tgp`, when operating with dense poolers, you can set `batched=False` to operate on sparse adjacency inputs (without dense padding/densification) and compute the auxiliary losses in that sparse mode.

Repo: https://github.com/tgp-team/torch-geometric-pool  
Docs (MinCutPooling): https://torch-geometric-pool.readthedocs.io/en/latest/api/poolers.html#tgp.poolers.MinCutPooling
